From dd579b720108bf24748a769e107ae246f7e32755 Mon Sep 17 00:00:00 2001
From: Robert Chiras <robert.chiras@nxp.com>
Date: Fri, 8 Dec 2017 15:15:35 +0200
Subject: [PATCH 3058/5242] MLK-17137-2: drm/bridge: Fix bridge_detach for
 nwl-dsi

commit  17b8663f670931781a769f79688494d9a380ef08 from
https://source.codeaurora.org/external/imx/linux-imx.git

When the bridge is detached from it's parent, we also need to unregister
the dsi_host. Also, in enable, check if we have a panel or a bridge
connected, otherwise enable is not needed.

Signed-off-by: Robert Chiras <robert.chiras@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/gpu/drm/bridge/nwl-dsi.c |    3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/drivers/gpu/drm/bridge/nwl-dsi.c b/drivers/gpu/drm/bridge/nwl-dsi.c
index fe64228..e10c0df 100644
--- a/drivers/gpu/drm/bridge/nwl-dsi.c
+++ b/drivers/gpu/drm/bridge/nwl-dsi.c
@@ -447,7 +447,7 @@ static void nwl_dsi_bridge_enable(struct drm_bridge *bridge)
 	struct device *dev = dsi->dev;
 	int ret;
 
-	if (dsi->enabled)
+	if (dsi->enabled || (!dsi->panel && !dsi->next_bridge))
 		return;
 
 	if (!dsi->lanes) {
@@ -1072,6 +1072,7 @@ static void nwl_dsi_bridge_detach(struct drm_bridge *bridge)
 	} else if (dsi->next_bridge) {
 		nwl_dsi_del_bridge(dsi->next_bridge->encoder, dsi->next_bridge);
 	}
+	mipi_dsi_host_unregister(&dsi->host);
 }
 
 static const struct drm_bridge_funcs nwl_dsi_bridge_funcs = {
-- 
1.7.9.5

